capsules/src/gpio_async.rs

```diff
@@ -0,0 +1,129 @@
+//! Provide userspace applications with a driver interface to asynchronous
+//! GPIO pins. These are pins that exist on something like a GPIO extender or
+//! a radio that has controllable GPIOs.
+
+use core::cell::Cell;
+use kernel::{AppId, Callback, Driver};
+use kernel::ReturnCode;
+
+use kernel::hil;
+
+pub struct GPIOAsync<'a, Port: hil::gpio_async::Port + 'a> {
+    ports: &'a [&'a Port],
+    callback: Cell<Option<Callback>>,
+    interrupt_callback: Cell<Option<Callback>>,
+}
+
+impl<'a, Port: hil::gpio_async::Port> GPIOAsync<'a, Port> {
+    pub fn new(ports: &'a [&'a Port]) -> GPIOAsync<'a, Port> {
+        GPIOAsync {
+            ports: ports,
+            callback: Cell::new(None),
+            interrupt_callback: Cell::new(None),
+        }
+    }
+
+    fn configure_input_pin(&self, port: usize, pin: usize, config: usize) -> ReturnCode {
+        let ports = self.ports.as_ref();
+        if config > 2 {
+            return ReturnCode::EINVAL;
+        }
+        let mode = match config {
+            0 => hil::gpio::InputMode::PullUp,
+            1 => hil::gpio::InputMode::PullDown,
+            _ => hil::gpio::InputMode::PullNone,
```

Comment:
I think the way this is written is a little confusing as it's not obvious that only 2 is valid for PullNone, I just checked and this is also valid Rust:
```diff
     fn configure_input_pin(&self, port: usize, pin: usize, config: usize) -> ReturnCode {
         let ports = self.ports.as_ref();
-        if config > 2 {
-            return ReturnCode::EINVAL;
-        }
         let mode = match config {
             0 => hil::gpio::InputMode::PullUp,
             1 => hil::gpio::InputMode::PullDown,
-            _ => hil::gpio::InputMode::PullNone,
+            2 => hil::gpio::InputMode::PullNone,
+            _ => return ReturnCode::EINVAL,
         };
         ports[port].make_input(pin, mode)
     }
```